### PR TITLE
Monitoring

### PR DESCRIPTION
**Pull request by alessio-creo**
needs #44 

Per deployment aggregations and the Cloudwatch alarms involved in tech debt

### CODE REVIEW

File: src/main/16-cloudwatch-alarms.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  name           = "${var.app_name}-error-logs-filter-${var.env}"
+  pattern        = "ERROR"
+  log_group_name = "/aws/eks/interop-probing-eks-dev/application"
+
+  metric_transformation {
+    name      = "ErrorCount"
+    namespace = "ApplicationLogs"
+    value     = 1
+  }
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:25]: Just a suggestion to organize metrics
```suggestion
  metric_transformation {
    name      = "ErrorCount"
    namespace = "EKSApplicationLogsFilters"
    value     = "1"

    dimensions = {
      PodApp = "$.pod_app"
    }
  }
```

File: src/main/16-cloudwatch-alarms.tf [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+resource "aws_cloudwatch_metric_alarm" "timestream_errors" {
+  alarm_name          = "${var.app_name}-cwalarm-timestream-system-errors-${var.env}"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = 10
+  metric_name         = "SystemErrors"
+  namespace           = "AWS/Timestream"
+  period              = 60
+  statistic           = "SampleCount"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:84]:
```suggestion
  statistic           = "Sum"
```

File: src/main/16-cloudwatch-alarms.tf [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+resource "aws_cloudwatch_metric_alarm" "lambda_errors" {
+  alarm_name          = "${var.app_name}-cwalarm-lambda-errors-${var.env}"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = 10
+  metric_name         = "Errors"
+  namespace           = "AWS/Lambda"
+  period              = 60
+  statistic           = "SampleCount"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:60]:
```suggestion
  statistic           = "Sum"
```

File: src/main/16-cloudwatch-alarms.tf [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+resource "aws_cloudwatch_metric_alarm" "error_logs" {
+  alarm_name          = "${var.app_name}-cwalarm-application-errors-${var.env}"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = 10
+  metric_name         = "ErrorCount"
+  namespace           = "ApplicationLogs"
+  period              = 60
+  statistic           = "SampleCount"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:35]:
```suggestion
  statistic           = "Sum"
```

File: src/main/15-sns-topic.tf
Comment (micdes-pagopa) [inline src/main/15-sns-topic.tf:1]: Missing topic policy to allow CloudWatch alarms to send notifications

File: src/main/16-cloudwatch-alarms.tf
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/16-cloudwatch-alarms.tf:1]: - ok
- ok
- Not in this PR but in #49 . I'll check and convert to draft that one

File: src/main/16-cloudwatch-alarms.tf (hunk 1 above)
Comment (alessio-creo) [reply src/main/16-cloudwatch-alarms.tf]: dimensions not supported

File: src/main/16-cloudwatch-alarms.tf [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  metric_transformation {
+    name      = "ErrorCount"
+    namespace = "EKSApplicationLogsFilters"
+    value     = "1"
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "error_logs" {
+  alarm_name          = "${var.app_name}-application-errors-${var.env}"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:97]: If you use the metric dimension I proposed (PodApp) you can create an alarm for each microservice, let me know if there are issues. The dimension works as expected on other projects

File: src/main/16-cloudwatch-alarms.tf [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    QueueName = each.value
+  }
+  threshold     = var.cw_alarm_thresholds.sqs_message_age
+  alarm_actions = [aws_sns_topic.cw_alarms.arn]
+}
+
+resource "aws_cloudwatch_metric_alarm" "lambda_errors" {
+  for_each            = toset(local.lambda_functions)
+  treat_missing_data  = "notBreaching"
+  alarm_name          = "${var.app_name}-lambda-errors-${each.value}-${var.env}"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:40]:
```suggestion
  alarm_name          = "${var.app_name}-${each.value}-errors-${var.env}"
```

File: src/main/16-cloudwatch-alarms.tf [hunk 7]
```diff
@@ -84,16 +84,12 @@ resource "aws_cloudwatch_log_metric_filter" "error_logs" {
...
+  log_group_name = "/aws/eks/${module.eks.cluster_name}/application"
 
   metric_transformation {
     name      = "ErrorCount"
     namespace = "EKSApplicationLogsFilters"
     value     = "1"
-
-    dimensions = {
-      PodApp = "$.pod_app"
-    }
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/16-cloudwatch-alarms.tf]: Yes. "Dimensions not supported"

[the remark this answers is not in the document]
Comment (micdes-pagopa) [reply src/main/16-cloudwatch-alarms.tf]: Can you double check? I use it in another project and it works 🤔 
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_log_metric_filter

File: src/main/16-cloudwatch-alarms.tf [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  metric_query {
+    id          = "sq"
+    expression  = "SERVICE_QUOTA(conc_util)"
+    return_data = true
+    label       = "LambdaConcurrencyUtilization"
+  }
+
+  metric_query {
+    id          = "conc_util"
+    return_data = true
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:72]:
```suggestion
    return_data = false
```
Only one metric query can return data

File: src/main/16-cloudwatch-alarms.tf [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  statistic           = "Sum"
+
+  dimensions = {
+    FunctionName = each.value
+  }
+  threshold     = 1
+  alarm_actions = [aws_sns_topic.cw_alarms.arn]
+}
+
+resource "aws_cloudwatch_metric_alarm" "lambda_concurrency_pct" {
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:55]: Missing threshold (60%) and datapoints to alarm (1)

File: src/main/16-cloudwatch-alarms.tf [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  treat_missing_data  = "notBreaching"
+  alarm_name          = "${var.app_name}-lambda-conc-util-pct-${var.env}"
+  comparison_operator = "GreaterThanUpperThreshold"
+  evaluation_periods  = 1
+  threshold_metric_id = "sq"
+  alarm_actions       = [aws_sns_topic.cw_alarms.arn]
+
+  metric_query {
+    id          = "sq"
+    expression  = "SERVICE_QUOTA(conc_util)"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:65]: We need to calculate the percentage of this, the alarm will trigger on 60% of the service quota

File: src/main/16-cloudwatch-alarms.tf [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    FunctionName = each.value
+  }
+  threshold     = 1
+  alarm_actions = [aws_sns_topic.cw_alarms.arn]
+}
+
+resource "aws_cloudwatch_metric_alarm" "lambda_concurrency_pct" {
+  treat_missing_data  = "notBreaching"
+  alarm_name          = "${var.app_name}-lambda-conc-util-pct-${var.env}"
+  comparison_operator = "GreaterThanUpperThreshold"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:58]:
```suggestion
  comparison_operator = "GreaterThanOrEqualToThreshold"
```

File: src/main/16-cloudwatch-alarms.tf (hunk 5 above)
Comment (alessio-creo) [reply src/main/16-cloudwatch-alarms.tf:117]: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html#extract-log-event-values

Seems like dimensions must match one of the specified in the metric filter

File: src/main/16-cloudwatch-alarms.tf [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+      PodApp = "$.pod_app"
+    }
+
+  }
+
+}
+
+resource "aws_cloudwatch_metric_alarm" "error_logs" {
+  for_each            = toset(local.microservices)
+  alarm_name          = "${var.app_name}-application-errors-${each.value}-${var.env}"
```
Comment (micdes-pagopa) [inline src/main/16-cloudwatch-alarms.tf:118]:
```suggestion
  alarm_name          = "${var.app_name}-${each.value}-app-errors-${var.env}"
```

**Pull request merged**